src/main/java/com/exasol/sql/expression/rendering/ValueExpressionRenderer.java

```diff
@@ -1,127 +1,239 @@
+    }
+
+    @Override
+    public void visit(final Function function) {
+        function.accept((FunctionVisitor) this);
+    }
+
+    @Override
+    public void visit(final BooleanExpression booleanExpression) {
+        booleanExpression.accept((BooleanExpressionVisitor) this);
     }
 
     @Override
     public void visit(final UnnamedPlaceholder unnamedPlaceholder) {
         append("?");
-        setLastVisited(unnamedPlaceholder);
     }
 
     @Override
     public void visit(final DefaultValue defaultValue) {
         appendKeyword("DEFAULT");
-        setLastVisited(defaultValue);
     }
 
+    /** Literal visitor **/
+    @Override
+    public void visit(final StringLiteral literal) {
+        append("'");
+        append(literal.toString());
+        append("'");
+    }
+
+    @Override
+    public void visit(final IntegerLiteral literal) {
+        append(literal.toString());
+    }
+
+    @Override
+    public void visit(final LongLiteral literal) {
+        append(literal.toString());
+    }
+
+    @Override
+    public void visit(final DoubleLiteral literal) {
+        append(literal.toString());
+    }
+
+    @Override
+    public void visit(final FloatLiteral literal) {
+        append(literal.toString());
+    }
+
+    @Override
+    public void visit(final BigDecimalLiteral literal) {
+        append(literal.toString());
+    }
+
+    @Override
+    public void visit(final NullLiteral nullLiteral) {
+        appendKeyword("NULL");
+    }
+
+    /** Function visitor */
+
     @Override
     public void visit(final ExasolFunction function) {
-        visitFunction(function);
+        renderFunction(function);
     }
 
-    private void visitFunction(final Function function) {
-        appendCommaWhenNeeded(function);
+    private void renderFunction(final AbstractFunction function) {
         appendKeyword(function.getFunctionName());
         if (function.hasParenthesis()) {
             startParenthesis();
         }
-    }
-
-    @Override
-    public void leave(final ExasolFunction function) {
+        ++this.nestedLevel;
+        this.visit(function.getParameters().toArray(ValueExpression[]::new));
+        --this.nestedLevel;
         if (function.hasParenthesis()) {
             endParenthesis();
         }
-        setLastVisited(function);
     }
 
     @Override
     public void visit(final ExasolUdf function) {
-        visitFunction(function);
+        renderFunction(function);
+        appendEmitsWhenNecessary(function);
     }
 
     @Override
-    public void leave(final ExasolUdf function) {
-        if (function.hasParenthesis()) {
-            endParenthesis();
-        }
-        appendEmitsWhenNecessary(function);
-        setLastVisited(function);
+    public void visit(final CastExasolFunction castFunction) {
+        appendKeyword("CAST");
+        startParenthesis();
+        ++this.nestedLevel;
+        castFunction.getValue().accept(this);
+        appendKeyword(" AS");
```

Comment:
Is the space after `AS` missing here on purpose? Because of parentheses maybe?

Reply:
It removed it because in the test there were two spaces. After having a closer look I found out that `ColumnsDefinitionRenderer` adds space sometimes.
So I added the space here again, changed the test to two spaces for now, and open an [issue](https://github.com/exasol/sql-statement-builder/issues/102) to refactor `ColumnsDefinitionRenderer`.